generator/statistics.cpp

```diff
@@ -53,22 +55,55 @@ namespace stats
 
     void operator() (FeatureType & f, uint32_t)
     {
-      f.ParseBeforeStatistic();
+      f.ParseHeader2();
 
-      FeatureType::InnerGeomStat const innerStats = f.GetInnerStatistic();
+      FeatureType::InnerGeomStat const innerStats = f.GetInnerStats();
 
       m_info.m_inner[0].Add(innerStats.m_points);
       m_info.m_inner[1].Add(innerStats.m_strips);
       m_info.m_inner[2].Add(innerStats.m_size);
 
-      // get geometry size for the best geometry
-      FeatureType::GeomStat const geom = f.GetGeometrySize(FeatureType::BEST_GEOMETRY);
-      FeatureType::GeomStat const trg = f.GetTrianglesSize(FeatureType::BEST_GEOMETRY);
+      // Get size stats and load the best geometry.
+      FeatureType::GeomStat const geom = f.GetOuterGeometryStats();
+      FeatureType::GeomStat const trg = f.GetOuterTrianglesStats();
 
-      m_info.m_byPointsCount[CountType(geom.m_count)].Add(geom.m_size);
-      m_info.m_byTrgCount[CountType(trg.m_count / 3)].Add(trg.m_size);
+      uint32_t outerGeomSize = 0, outerTrgSize = 0;
+      for (size_t ind = 0; ind < DataHeader::kMaxScalesCount; ++ind)
+      {
+        auto const geomSize = geom.m_sizes[ind], geomElems = geom.m_elements[ind],
+                   trgSize = trg.m_sizes[ind], trgElems = trg.m_elements[ind];
+        m_info.m_byLineGeom[ind].Add(geomSize, geomElems);
+        outerGeomSize += geomSize;
+        m_info.m_byAreaGeom[ind].Add(trgSize, trgElems);
+        outerTrgSize += trgSize;
+
+        if (ind > 0)
+        {
+          // If a feature has a more simplified version of current geometry.
+          if (geom.m_elements[ind - 1] > 0)
+            m_info.m_byLineGeomCompared[ind].Add(geomSize, geomElems);
+          if (trg.m_elements[ind - 1] > 0)
+            m_info.m_byAreaGeomCompared[ind].Add(trgSize, trgElems);
+        }
+
+        if (ind < DataHeader::kMaxScalesCount - 1)
+        {
+          // If feature's current geometry almost duplicates a more detailed one
+          // (has <geometryDupFactor less elements).
+          if (geomElems > 0 && geomElems * m_info.m_geometryDupFactor > geom.m_elements[ind + 1])
+            m_info.m_byLineGeomDup[ind].Add(geomSize, geomElems);
+          if (trgElems > 0 && trgElems * m_info.m_geometryDupFactor > trg.m_elements[ind + 1])
+            m_info.m_byAreaGeomDup[ind].Add(trgSize, trgElems);
```

Comment:
geomElems > 0, trgElems > 0 are useless here

Reply:
you're right, thanks!